pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsGenerator.java

```diff
@@ -113,6 +117,151 @@ public void release() {
         }
     }
 
+    /**
+     * A wraps the response buffer and asynchronously provides a gzip compressed buffer when requested.
+     */
+    public static class ResponseBuffer {
+        private final ByteBuf uncompressedBuffer;
+        private boolean released = false;
+        private CompletableFuture<ByteBuf> compressedBuffer;
+
+        private ResponseBuffer(final ByteBuf uncompressedBuffer) {
+            this.uncompressedBuffer = uncompressedBuffer;
+        }
+
+        public ByteBuf getUncompressedBuffer() {
+            return uncompressedBuffer;
+        }
+
+        public synchronized CompletableFuture<ByteBuf> getCompressedBuffer(Executor executor) {
+            if (released) {
+                throw new IllegalStateException("Already released!");
+            }
+            if (compressedBuffer == null) {
+                compressedBuffer = new CompletableFuture<>();
+                ByteBuf retainedDuplicate = uncompressedBuffer.retainedDuplicate();
+                executor.execute(() -> {
+                    try {
+                        compressedBuffer.complete(compress(retainedDuplicate));
+                    } catch (Exception e) {
+                        compressedBuffer.completeExceptionally(e);
+                    } finally {
+                        retainedDuplicate.release();
```

Comment:
It seems we don't release `uncompressedBuffer ` after it compressed?

Reply:
It's handled in the release method later in the same ResponseBuffer inner class. The release method gets called when the reference count of the MetricsBuffer class reaches 0. It will release both the uncompressed and possible compressed buffer at that point.

Reply:
@lhotari Yes, I've noticed.
I mean we can release `uncompressedBuffer` immediately after it compressed, and then, only one of them(`uncompressedBuffer ` and `compressedBuffer `) stay in the memory at the same time. It may reduce the memory usage, WDYT?

Reply:
@dao-jun that's a potential further optimization. However, the memory will be held for a rather short time. The benefit of the current solution in this PR is that it will serve concurrent requests where there's a mix of ordinary http requests and http requests with `Accept-Encoding: gzip`. Optimizing further will add more complexity without much gains.

Reply:
OK, I got it